---
title:   "<img id=\"logo\" style=\"width: 100px;\" src=\"figures/logo.jpg\" /> QTL mapping with ovary counts from DO mice"
author: "Selcan Aydin"
date: "`r Sys.Date()`"
output: 
 html_document:
    code_download: true
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
    includes:
      after_body: include_footer.html
---


```{r setup, warning=FALSE, message=FALSE, results='hide'}

## set chunk options
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress = FALSE)

## load libraries
library(tidyverse)
library(qtl2)
library(here)
library(assertthat)
library(readxl)
library(ggpubr)
source(here("functions.R") )#for qtl2 split_map function and others

# graphical parameters
# founder_colors <- c("#FFDC00", "#888888", "#F08080", "#0064C9", "#7FDBFF", "#2ECC40", "#FF4136", "#B10DC9")
# founder colors - color blind friendly pallette
founder_colors <- c(AJ = "#F0E442", B6 = "#555555", `129` = "#E69F00", NOD = "#0072B2",
   NZO = "#56B4E9", CAST = "#009E73", PWK = "#D55E00", WSB = "#CC79A7")
        
```


```{r read_data}

## read in covariates:
### Ruby shared 3 files: DO_DOB (date of birth), DO_date of collection, DO_Weights 
dob <- readxl::read_xlsx(here("data","DO_DOB.xlsx"))
doc <- readxl::read_xlsx(here("data","DO_date of collection.xlsx"))
weights <- readxl::read_xlsx(here("data","DO_Weights.xlsx"))

### merge covariates
covar_mat_new <- dob %>% 
  mutate( DOB = as.factor(DOB)) %>% 
  full_join( doc %>% 
               rename( DOC = `Collection date`) %>% 
               mutate( DOC = as.factor(DOC))
             ) %>% 
  full_join( weights %>% 
               rename( Name = `Name *`) %>% 
               mutate( Weight = as.double(Weight))
             ) %>% 
  mutate( animal_id = gsub("-Q-","_QTL_",Name))

### read in the previous covariate table to add sex, generation, histology
### correcting sample names -- to match the phenotype matrix
covar_mat <- read_csv( here("data","GM_covar_whistology.csv")) %>% 
  mutate( generation = factor(generation)) %>% 
  mutate( animal_id = gsub("-","_",id)) %>% # need to change the - to _ first then remove the _T
  mutate( animal_id = gsub("_T","",animal_id)) # need to change the - to _ first then remove the _T
  

### merge old and new covariates
covar_mat_extended <- covar_mat_new %>% 
  full_join( covar_mat)

### make the covar matrix using generation
covar<- model.matrix( ~generation, data = covar_mat_extended)
covar <- covar[,-1] #removing intercept column
rownames(covar) <- covar_mat_extended$animal_id

## read in genotype probabilities
probs <- readRDS( here("data","Filas_DO_MURGIGV01_20190131_20200220__GigaMUGA_genoprobs_8state_69k_sorted.rds")) 

## read in marker details and make gmap & pmap
#### gmap
mapfile <- (here("data","ref.genome_grid_69k.for_qtl2.csv"))
markers <- dimnames(probs)[[3]]
map_dat <- qtl2::read_csv(mapfile)[markers, ]
# assert_that(are_equal(rownames(map_dat), markers))
gmap <- split_map(map_dat)
map_dat$marker <- rownames(map_dat)
#### pmap
map_dat2 <- map_dat %>%
  separate(marker, into=c('chrom', 'pos_bp'), convert=T, remove=F) %>%
  mutate(n=1:n()) %>% 
  as_tibble()
pmap <- split_map(dplyr::select(map_dat2, marker,
                                chr, pos_bp) %>% as.data.frame() %>%
                    tibble::remove_rownames() %>%
                    tibble::column_to_rownames('marker'))

##  genoprobs need to be modified to match qtl2 formatting
####  This function is from Dan Skelly:
message("converting probs to qtl2 and calculating kinship matrix")
uchroms <- unique(map_dat$chr)
probs_3d_to_qtl2 <- function(genoprobs) {
  # Convert to qtl2 genoprobs format
  # Similar to qtl2convert::probs_doqtl_to_qtl2()
  markers <- dimnames(genoprobs)[[3]]
  chroms <- sapply(strsplit(markers, "_"), "[[", 1)
  newprobs <- vector("list", length(uchroms))
  names(newprobs) <- uchroms
  for (chrom in uchroms) newprobs[[chrom]] <- genoprobs[, , chroms == chrom]
  attr(newprobs, "is_x_chr") <- c(rep(FALSE, length(uchroms)-1),TRUE)
  attr(newprobs, "crosstype") <- "DO"
  attr(newprobs, "alleles") <- c("A","B","C","D","E","F","G","H")
  attr(newprobs, "alleleprobs") <- TRUE
  class(newprobs) <- c("calc_genoprob", "list")
  newprobs
}
genoprobs <- probs_3d_to_qtl2(probs)
## subsetting and correcting the sample names -- to match the phenotype matrix
genoprobs <- genoprobs[ind = covar_mat_extended$id]
new.ids <- covar_mat_extended$animal_id
names(new.ids) <- covar_mat_extended$id
genoprobs <- replace_ids(genoprobs, ids = new.ids)


##  prep some stuff for plotting:
uchr <- c(as.character(1:19), "X")
cl <- dplyr::select(map_dat2, chr, pos_bp) %>%
  group_by(chr) %>%
  dplyr::summarize(len = max(pos_bp))
clp <- with(cl, setNames(len, chr))
chrom_lens <- setNames(as.numeric(clp[uchr]), uchr)
chrom_lens_offset <- cumsum(chrom_lens) - chrom_lens
chrom_lens_midpt <- chrom_lens_offset + chrom_lens / 2
map_dat2$pos_bp_offset <- map_dat2$pos_bp+chrom_lens_offset[map_dat2$chr]


## read in phenotype data
ovary_counts <- readxl::read_xlsx(here("data","DO_Pheno.xlsx"), 
                                  skip = 1,
                           col_names = c("animal_id",
                                         "ovary1_pos1",
                                         "ovary2_pos1",
                                         "sum_pos1"
                                         )# override column names
                           ) %>% 
  column_to_rownames("animal_id")

### annotate the ovary counts + normalize + avg. etc. into a tibble.
ovary_counts_pos_long <- ovary_counts %>% 
  rownames_to_column("animal_id") %>% 
  select(c("animal_id",
           "ovary1_pos1","ovary2_pos1")
         ) %>% # remove the sum + average, will remake them
  pivot_longer(cols = 2:3, 
               names_to = c("ovary","position"), names_sep = "_", 
               values_to = "count"
  ) %>% 
  mutate( ovary = factor(str_sub(ovary,6)),
          position = factor(str_sub(position,4))
          ) 

### have the phenotypes in a matrix format for PCA later
ovary_counts_pos_long %>% 
  pivot_wider( id_cols = "animal_id", 
               values_from = "count",
               names_from = c("ovary","position"),
               names_glue = "ovary{ovary}_position{position}"
               ) %>% 
  column_to_rownames("animal_id")-> ovary_counts_pos_wide

### Load in coat color to map as positive control
coat_color <- readxl::read_xlsx(here("data","DO_coat color.xlsx") ) %>% 
  mutate( animal_id = gsub("-Q-","_QTL_",Name) ) %>% 
  select(-Name) %>% 
  mutate( albino = case_when( `Coat Color` == "black" ~ 0, 
                            `Coat Color` == "Black" ~ 0, 
                            `Coat Color` == "Albino" ~ 1,
                            `Coat Color` == "albino" ~ 1,
                            `Coat Color` == "Agouti" ~ 0,
                            `Coat Color` == "agouti" ~ 0),
          black = case_when( `Coat Color` == "black" ~ 1, 
                            `Coat Color` == "Black" ~ 1, 
                            `Coat Color` == "Albino" ~ 0,
                            `Coat Color` == "albino" ~ 0,
                            `Coat Color` == "Agouti" ~ 0,
                            `Coat Color` == "agouti" ~ 0)
          ) %>% 
  column_to_rownames("animal_id")

## subset the data to only include the animals with phenotype
### get samples
sample_ids <- unique(ovary_counts_pos_long$animal_id)
### subset genoprobs then calculate kinship
genoprobs <- genoprobs[ ind = sample_ids]
### get kinship matrix for mapping
kinship_loco <- qtl2::calc_kinship(genoprobs, "loco", cores=1)
### subset covar
covar <- covar[sample_ids,,drop=FALSE]

  
## controls for mapping -- check that the sample names are matching across the data sets we are using for mapping.
# assert_that(are_equal(rownames(genoprobs[[1]]), rownames(ovary_counts_pos_wide_rankZ)))
# assert_that(are_equal(rownames(genoprobs[[1]]), rownames(covar)))

```

This is the overview, processing and QTL mapping analysis of ovary counts obtained from DO mice by Ruby Boateng. The genotype probabilities were provided by Belinda Cornes in CS. You can download the code using the link on the upper right corner or hide/view chunks of interest or check out the [github repo](https://github.com/selcant/DO_ovary_qtl). The full list of R packages can be found in the [renv.lock](https://github.com/selcant/DO_ovary_qtl/blob/main/renv.lock) file. 

## Overview of the data {.tabset .tabset-fade .tabset-pills}


### Distribution of ovary counts across DO animals. 

```{r ova_dist_plot, fig.width=8, fig.height=4}

ovary_counts_pos_long %>% 
  mutate( ovary = paste("ovary",ovary),
          position = paste("position", position)) %>% 
  ggplot()+
  aes( x = count)+
  #geom_density()+
  geom_histogram( )+
  theme_pubclean(base_size = 16)+
  facet_wrap(~ovary+position)

```

### Counts don't show significant differences across two ovaries.

```{r ovary_count_visual, fig.width=4, fig.height=4, fig.cap = "Box plot comparing the counts from ovariy 1 and 2."}

ovary_counts_pos_long %>% 
  left_join( covar_mat) %>% 
  ggplot()+
  aes(y = count,
      col = ovary,
      x = ovary)+
  geom_boxplot( width = 0.1)+
  ylab("Ovary count")+
  xlab("Ovary #")+
  labs(col = "Ovary #")+
  stat_compare_means()+ # wilcoxon, p = 0.5
  theme_pubclean(base_size = 16)+
  color_palette("npg")-> ova_plot

# ovary_counts_pos_long %>% 
#   left_join( covar_mat) %>% 
#   ggplot()+
#   aes(y = count,
#       col = ovary,
#       x = position)+
#   geom_boxplot( width = 0.2)+
#   ylab("Ovary count")+
#   xlab("Slide position")+
#   labs(col = "Ovary #")+
#   #stat_compare_means()+ # kruskall-wallis, all p > 0.5
#   theme_pubclean(base_size = 16)+
#   color_palette("npg") -> ova_pos_plot

ggarrange(ova_plot, common.legend = TRUE, nrow = 1)

```

### Ovary counts do not show clear bias across technical covariates. 

```{r cor_plots, fig.width=8, fig.height=10}

ovary_counts_pos_long %>% 
  left_join(covar_mat_extended) %>% 
  ggplot()+
  aes( x = count,
       y = Weight)+
  geom_point()+
  facet_wrap(~ovary+position)+
  theme_pubclean(base_size = 16)+
  stat_smooth(method = "lm")  +
  xlab("Ovary counts")+
  ylab("Animal weight")-> weight_plot

ovary_counts_pos_long %>% 
  left_join(covar_mat_extended) %>% 
  ggplot()+
  aes( x = count,
       y = DOB)+
  geom_point()+
  facet_wrap(~ovary+position)+
  theme_pubclean(base_size = 16) +
  xlab("Ovary counts")+
  ylab("Date of birth")-> dob_plot


ovary_counts_pos_long %>% 
  left_join(covar_mat_extended) %>% 
  ggplot()+
  aes( x = count,
       y = DOC)+
  geom_point()+
  facet_wrap(~ovary+position)+
  theme_pubclean(base_size = 16)  +
  xlab("Ovary counts")+
  ylab("Date of collection")-> doc_plot


ggarrange( ggarrange(NULL,weight_plot, nrow =1, widths = c(0.15,1)), dob_plot, doc_plot, nrow = 3)
```


<!-- ### First two principal components (PCs) explain more than 99% of the variation with partial grouping based on technical covariates. -->

<!-- ```{r pca_plot, fig.width=16,fig.height=12, fig.cap="On top, scree plot showing the percent variation explained by each principal component. On bottom, scatterplot of PC1 and PC2 values where the samples are colored by technical covariates DO generation and batch corresponding to the date of sample processing by histology."} -->

<!-- pca_ova <- prcomp( (ovary_counts_pos_wide), center = T, scale. = T) -->
<!-- percentVar <- round(100 * pca_ova$sdev^2 / sum(pca_ova$sdev^2),2) -->
<!-- pca_ova_df <- pca_ova$x %>%  -->
<!--   as.data.frame() %>%  -->
<!--   rownames_to_column("animal_id") %>%  -->
<!--   left_join( covar_mat_extended ) -->

<!-- tibble( var_exp = percentVar, -->
<!--         num = seq(1:2)) %>% -->
<!--   ggplot()+ -->
<!--   aes( x = as.factor(num), y = var_exp)+ -->
<!--   geom_bar( stat = "identity")+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   xlab("PC #")+ -->
<!--   ylab("% Variation explained")+ -->
<!--   ylim(0,100)-> ova_pca_scree_plot -->

<!-- pca_ova_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = generation)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",percentVar[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",percentVar[2],"%)"))+ -->
<!--   labs(col="DO Generation")-> ova_pca_plot_gen -->

<!-- pca_ova_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = histology_batch)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",percentVar[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",percentVar[2],"%)"))+ -->
<!--   labs(col="Batch  ")-> ova_pca_plot_batch -->

<!-- pca_ova_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = DOB)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",percentVar[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",percentVar[2],"%)"))+ -->
<!--   labs(col="DOB")-> ova_pca_plot_dob -->

<!-- pca_ova_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = DOC)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",percentVar[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",percentVar[2],"%)"))+ -->
<!--   labs(col="Date of Collection")-> ova_pca_plot_doc -->

<!-- pca_ova_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = Weight)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   scale_color_gradient()+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",percentVar[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",percentVar[2],"%)"))+ -->
<!--   labs(col="Animal Weight")-> ova_pca_plot_w -->

<!-- ggarrange(ova_pca_scree_plot, ova_pca_plot_gen,ova_pca_plot_batch,  -->
<!--           ova_pca_plot_dob, ova_pca_plot_doc, ova_pca_plot_w, nrow = 3, ncol = 2) -->

<!-- ``` -->


<!-- ### Regressing out PC1 to remove variation introduced by technical covariates.  -->

<!-- ```{r regress_pc1} -->


<!-- ova_pc1_reg <- apply( ovary_counts_pos_wide, 2, function(y){ -->
<!--                                                 lm_obj=lm(y~pca_ova_df$PC1); -->
<!--                                                 lm_obj$residuals} -->
<!--   ) -->

<!-- # check how residuals look -->
<!-- # apply( ovary_counts_pos_wide, 2, function(y){ -->
<!-- #                                                 lm_obj=lm(y~pca_ova_df$PC1); -->
<!-- #                                                 hist(lm_obj$residuals)} -->
<!-- #   ) -->


<!-- ``` -->


<!-- ```{r new_pca_plot, fig.width=10,fig.height=8, fig.cap="On top, scree plot showing the percent variation explained by each principal component. On bottom, scatterplot of PC1 and PC2 values where the samples are colored by technical covariates DO generation and batch corresponding to the date of sample processing by histology."} -->

<!-- pca_ova_pc1_reg <- prcomp(ova_pc1_reg) -->
<!-- var_exp <- round(100 * pca_ova_pc1_reg$sdev^2 / sum(pca_ova_pc1_reg$sdev^2),2) -->
<!-- pca_ova_pc1_reg_df <- pca_ova_pc1_reg$x %>%  -->
<!--   as.data.frame() %>%  -->
<!--   rownames_to_column("animal_id") %>%  -->
<!--   left_join( covar_mat ) -->

<!-- tibble( var_exp = var_exp, -->
<!--         num = seq(1:10)) %>% -->
<!--   ggplot()+ -->
<!--   aes( x = as.factor(num), y = var_exp)+ -->
<!--   geom_bar( stat = "identity")+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   xlab("PC #")+ -->
<!--   ylab("% Variation explained")+ -->
<!--   ylim(0,100)-> ova_pca_scree_plot_pc1_reg -->

<!-- pca_ova_pc1_reg_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = generation)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",var_exp[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",var_exp[2],"%)"))+ -->
<!--   labs(col="DO Generation")-> ova_pca_plot_gen_pc1_reg -->

<!-- pca_ova_pc1_reg_df %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = PC1, -->
<!--       y = PC2, -->
<!--       col = histology_batch)+ -->
<!--   geom_point( size = 3, alpha = 0.7)+ -->
<!--   theme_pubclean(base_size = 16)+ -->
<!--   color_palette("Paired")+ -->
<!--   theme(legend.position = "top")+ -->
<!--   xlab(paste0("PC1 (",var_exp[1],"%)"))+ -->
<!--   ylab(paste0("PC2 (",var_exp[2],"%)"))+ -->
<!--   labs(col="Batch  ")-> ova_pca_plot_batch_pc1_reg -->


<!-- ggarrange( ggarrange(NULL, ova_pca_scree_plot_pc1_reg, NULL, nrow = 1, widths = c(0.2,0.6,0.2)),   -->
<!--           ggarrange(ova_pca_plot_gen_pc1_reg,ova_pca_plot_batch_pc1_reg, nrow = 1), nrow = 2, heights = c(0.6, 1)) -->



<!-- ``` -->

## Genetic mapping with coat color as a control identifies known loci associated with albino and black coat color with high LOD scores, suggesting we have reasonable mapping power in the data. {.tabset .tabset-fade .tabset-pills}

### Heritability estimates.

```{r estimate_heritability}

# heritability
full_shared_kinship <- calc_kinship( genoprobs, type="overall")

coat_color_herit <- est_herit(
  pheno = coat_color[ rownames(full_shared_kinship),c("albino","black"),drop=F],
  kinship = full_shared_kinship,
  addcovar = covar
) %>%
  as_tibble( rownames = "Phenotype") %>% 
  mutate( value = round(value, 2)) %>% 
  rename( "Heritability estimate"= "value" )

coat_color_herit

```
### QTL map

```{r qtl_mapping_coat_col, cache = TRUE}

# mapping for all the phenotypes
coat_col_scan <- qtl2::scan1(genoprobs, 
                       coat_color[ ,c("albino","black"),drop=F],
                       kinship_loco, 
                       addcovar= covar[sample_ids,,drop=FALSE], 
                       cores=1) 

qtl_eff_b <-scan1blup(genoprobs =genoprobs[,"2"],
                   pheno = coat_color[,"black",drop=F],
                   kinship = kinship_loco[["2"]],
                   addcovar = covar[sample_ids,,drop=FALSE]) 

qtl_eff_a <- scan1blup(genoprobs =genoprobs[,"7"],
                   pheno = coat_color[ ,"albino",drop=F],
                   kinship = kinship_loco[["7"]],
                   addcovar = covar[sample_ids,,drop=FALSE]) 

find_peaks(coat_col_scan,pmap, threshold = 10)

```

Albino coat QTL scan:

```{r albino_plot, fig.width=8, fig.height=6}

plot_coefCC( qtl_eff_a, scan1_output = coat_col_scan[,"albino",drop=F], pmap, main="Albino", legend = "topleft")

```


Black coat QTL scan:

```{r black_plot, fig.width=8, fig.height=6}

plot_coefCC( qtl_eff_b, scan1_output = coat_col_scan[,"black",drop=F], pmap, main="Black", legend= "topleft")

```


## Ovary counts are heritable and we identify a significant (alpha = 0.1) QTL on chromosome 17. {.tabset .tabset-fade .tabset-pills}


### Heritability estimates

```{r est_herit_ova}

# use rankZ normalized values for all three

ovary_counts_rankZ <- apply(ovary_counts, 2, rankZ)

ova_counts_herit <- est_herit(
  pheno = ovary_counts_rankZ,
  kinship = full_shared_kinship,
  #addcovar = covar
) %>%
  as_tibble( rownames = "Phenotype") %>% 
  mutate( value = round(value, 2)) %>% 
  rename( "Heritability estimate"= "value" )

ova_counts_herit

```


```{r qtl_mapping, cache = TRUE}

# mapping for all the phenotypes
ova_counts_qtl_scan <- qtl2::scan1(genoprobs, 
                       ovary_counts_rankZ,
                       kinship_loco, 
                       addcovar= covar[sample_ids,,drop=FALSE], 
                       cores=1) 

# use temporary threshold of 6
qtl_peaks <- find_peaks( ova_counts_qtl_scan, map = gmap, threshold = 5)



```


```{r qtl_threshold, eval = T}

## running a stand alone script for this 
## see at get_perm.R
## saving objesct for that script here
# save( genoprobs, rankz_all_pheno, kinship_loco, covar ,file = here("data","perm_objects.RData"))
# save(genoprobs, ovary_counts_rankZ, kinship_loco,covar, file = here("data", "perm_objects_v2.RData")) 

# loading the perm
load(here("data/perm_thres_v2.RData"))
# get 0.05, 0.1 and 0.5 thresholds for each phenotype
perm_thres_df <- qtl_map_perm_thres %>%
  as_tibble() %>%
  summarise_all(.funs = list(t_005 = ~ quantile(., 0.95),
                             t_01 = ~ quantile(., 0.9),
                             t_05 = ~ quantile(., 0.5))
                             ) %>%
  pivot_longer(cols = everything(), names_to = c("phenotype","threshold"), names_sep = "_t_", values_to = "value")


qtl_peaks <- qtl_peaks %>% 
  left_join( ., perm_thres_df, by = c("lodcolumn"="phenotype")) %>% 
  filter( lod >= value) %>% 
  mutate(peak_cM= pos, peak_chr = chr) %>% 
  interp_bp() 
#qtl_peaks_wo_covar <- find_peaks( qtl_map_wo_covar, map = gmap, threshold = 6)

# Get the bounding markers for each QTL peak
# i.e. markers on the 69k grid that are up- and downstream of the peak
query <- qtl_peaks %>% 
  dplyr::select(peak_chr, interp_bp_peak) %>%
  dplyr::rename(chrom=peak_chr, start=interp_bp_peak) %>% 
  mutate(end=start) %>%
  GenomicRanges::GRanges()
subject <- select(map_dat2, chrom, pos_bp) %>% 
  dplyr::rename(start=pos_bp) %>%
  mutate(end=start) %>% 
  GenomicRanges::GRanges()   # length 69,005

qtl_peaks$before <- map_dat2$marker[GenomicRanges::follow(query, subject)]
qtl_peaks$after <- map_dat2$marker[GenomicRanges::precede(query, subject)]

```


### QTL map of rankZ normalized DO ovary counts


```{r qtl_peaks_plot, fig.width=12, fig.height=6, fig.cap="Lineplot showing the genomic location of significant and suggestive QTL peaks for all phenotypes with LOD scores annotated. Genetic mapping is done using generation of DO animals as an additive covariate with ovary counts from 174 animals in total."}

par(mar=c(5,8,4,1))
plot_peaks( filter(qtl_peaks), gmap, lod_labels = TRUE, tick_height = 0.5)
title("QTL peaks of all phenotypes, mapped using\n DO generation as a covariate",cex.main=1.5)

```


```{r}

qtl_peaks %>% 
  mutate(threshold = as.numeric(threshold)/10) %>% 
  rename( Phenotype = lodcolumn, 
          `Peak (bp)` = interp_bp_peak,
          `Peak (cM)` = peak_cM,
          `LOD score` = lod,
          `Significance alpha` =threshold,
          `Significance LOD` = value
          ) %>% 
  mutate_if( is.numeric, round, 2) %>% 
  select( -lodindex) %>% 
  DT::datatable(.,
     extensions = 'Buttons',
                rownames = FALSE, 
                filter="top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel'),
                               pageLength = 10, 
                               scrollX= TRUE
                               ))
  

```


### QTL with allele effects {.tabset .tabset-fade .tabset-pills}

```{r allele_eff_plots, fig.height=6, fig.width=8,results='asis',echo=FALSE, eval=TRUE,cache=TRUE, warning=FALSE, message=FALSE}

all_df_effs <- c()
haps <- LETTERS[1:8]

qtl_peaks %>% 
  filter(  !(lodcolumn =="ovary1_pos1" & threshold == "05" )) -> qtl_peaks_lod6

for( i in 1:nrow(qtl_peaks_lod6)){
  chrom <- qtl_peaks_lod6$chr[i]
  pc <- ovary_counts_rankZ[,qtl_peaks_lod6$lodcolumn[i]]
  this_markers <- c(qtl_peaks_lod6$before[i], qtl_peaks_lod6$after[i])

  scan1 <- scan1(genoprobs = genoprobs,
                 pheno = pc,
                 kinship = kinship_loco,
                 addcovar = covar[sample_ids,,drop=FALSE])
  effs1 <- scan1blup(genoprobs =genoprobs[,chrom],
                   pheno = pc,
                   kinship = kinship_loco[[chrom]],
                   addcovar = covar[sample_ids,,drop=FALSE])


  all_df_effs[[i]] <- colMeans(effs1[this_markers, haps])
  cat('####',qtl_peaks_lod6$lodcolumn[i],'peak on chr ', chrom ,' \n')
  plot_coefCC(effs1, pmap, scan1_output = scan1, main=paste0(qtl_peaks_lod6$lodcolumn[i]," plot"))
  cat(' \n \n')

}

```

## Follow up on QTL on chromosome 17 {.tabset .tabset-fade .tabset-pills}

### Closer look at the QTL

```{r qtl_zoom_plot, fig.height=4, fig.width=10}

# conf_int_chr1 <- bayes_int( scan1_output = ova_counts_qtl_scan, map = pmap, lodcolumn ="ovary2_pos1", threshold = 6, chr = 1) %>%  as_tibble()
conf_int_chr17 <- bayes_int( scan1_output = ova_counts_qtl_scan, map = pmap, lodcolumn ="ovary1_pos1", threshold = 6, chr = 17) %>% as_tibble()

# ova_counts_qtl_scan %>% 
#   as_tibble( rownames = "marker") %>% 
#   select( marker, lod = sum_pos1) %>% 
#   left_join(map_dat2) %>% 
#   filter( chrom %in% c(1)) %>% 
#   ggplot()+
#     aes( 
#       x= pos_bp/1e06,
#       y = lod
#       )+
#     geom_line( size = 1)+
#     theme_pubclean( base_size = 18)+
#     facet_wrap(~chrom)+
#   xlab("Chr 1 (Mb)")+
#   ylab("LOD")+
#   xlim( conf_int_chr1$ci_lo/1e06, conf_int_chr1$ci_hi/1e06)+
#   geom_vline( xintercept = conf_int_chr1$pos/1e06, col = "dark red", linetype = 2, size = 1) -> chr1_plot


ova_counts_qtl_scan %>% 
  as_tibble( rownames = "marker") %>% 
  select( marker, lod = ovary1_pos1) %>% 
  left_join(map_dat2) %>% 
  filter( chrom %in% c(17)) %>% 
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod
      )+
    geom_line( size = 1)+
    theme_pubclean( base_size = 18)+
    #facet_wrap(~chrom)
  xlab("Chr 17 (Mb)")+
  ylab("LOD")+
  xlim( conf_int_chr17$ci_lo/1e06, 20+conf_int_chr17$pos/1e06)+
  geom_vline( xintercept = conf_int_chr17$pos/1e06, col = "dark red", linetype = 2, size = 1) -> chr17_plot

chr17_eff_plot <- all_df_effs[[2]] %>%
  as_tibble(rownames = "effect") %>% 
  mutate( pheno = "sum_pos1") %>%
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  ggplot()+
  aes( x = effect,
       y = value,
       group = pheno
       )+
  geom_point(size = 4, show.legend = F)+
  geom_line(show.legend = T, size =1.2)+
  theme_pubclean(base_size = 18)+
  ylab("Haplotype effects")+
  xlab("")+
  ylim(-1,1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  coord_flip( clip ="off")+
  theme(legend.position = "none")
  
  
ggarrange(  chr17_plot,chr17_eff_plot, widths = c(0.8,0.4) ,nrow =1)

```


### Genes under the QTL peak

List of genes that are within the +/- 5Mb of the QTL peak.

```{r gene_list}

# v102 genes, the last ensembl version with the old genome assembly
all_genes <- read_tsv( here( "data", "mouse_genes_v102.txt"), 
                       col_names = c("ensembl_gene_id", 
                                     "gene_start", "gene_end","strand",
                                     "gene_biotype","mgi_symbol", "gene_chr"),
                       skip = 1)

all_genes %>% 
  mutate( gene_midpoint = (gene_start+gene_end)/2) %>% 
  filter( 
    # (gene_chr == 1 & between( gene_midpoint, conf_int_chr1$pos-5e06, conf_int_chr1$pos+5e06) ) |
            (gene_chr == 17 & between( gene_midpoint, conf_int_chr17$pos-5e06, conf_int_chr17$pos+5e06) )
  ) %>% 
  mutate( qtl_peak = case_when( 
    #gene_chr ==1 ~conf_int_chr1$pos,
                                         gene_chr ==17~conf_int_chr17$pos)
          ) %>% 
  mutate( distance = abs(gene_midpoint - qtl_peak)) %>% 
  select(ensembl_gene_id, mgi_symbol, gene_chr, gene_midpoint, distance_to_qtl = distance, gene_biotype) %>% 
  arrange(gene_chr, (distance_to_qtl)) %>% 
   DT::datatable(.,
     extensions = 'Buttons',
                rownames = FALSE, 
                filter="top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel'),
                               pageLength = 10, 
                               scrollX= TRUE
                               ))

```

### Association mapping doesn't higlight any clear candidate SNPs.


```{r as_mapping}

## adding cc_variants for association mapping
query_variants <- create_variant_query_func(here("data/cc_variants.sqlite"))
query_genes <- create_gene_query_func("data/mouse_genes_mgi.sqlite")
pmap_Mbp <- lapply( pmap, function(x){x/1e06})
attributes(genoprobs)$is_x_chr <- NULL

# out_snps_chr1 <- scan1snps(genoprobs = genoprobs, 
#                            map = pmap_Mbp, 
#                            pheno = ovary_counts_rankZ[,"sum_pos1",drop=F], 
#                            kinship = kinship_loco[["1"]], 
#                            addcovar = covar, 
#                            query_func= query_variants,
#                            chr= 1, 
#                            start = (conf_int_chr1$pos/1e06)-2.5, 
#                            end = (conf_int_chr1$pos/1e06)+2.5, 
#                            keep_all_snps = F)
out_snps_chr17 <- scan1snps(genoprobs = genoprobs, 
                           map = pmap_Mbp, 
                           pheno = ovary_counts_rankZ[,"sum_pos1",drop=F], 
                           kinship = kinship_loco[["17"]], 
                           addcovar = covar, 
                           query_func= query_variants,
                           chr= 17, 
                           start = (conf_int_chr17$pos/1e06)-2.5, 
                           end = (conf_int_chr17$pos/1e06)+2.5, 
                           keep_all_snps = F)


```


```{r as_mapping_plot_chr1, fig.height=8, fig.width=10, eval = F}

genes_chr1 <- query_genes(chr = 1,
                          start = (conf_int_chr1$pos/1e06)-2.5, 
                          end = (conf_int_chr1$pos/1e06)+2.5)
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot(out_snps_chr1$lod, out_snps_chr1$snpinfo, drop_hilit=1.5, genes=genes_chr1)

```


```{r as_mapping_plot_chr17, fig.height=10, fig.width=14}

genes_chr17 <- query_genes(chr = 17,
                          start = (conf_int_chr17$pos/1e06)-2.5, 
                          end = (conf_int_chr17$pos/1e06)+2.5)
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot(out_snps_chr17$lod, out_snps_chr17$snpinfo, drop_hilit=1.5, genes=genes_chr17)

```


## Looking at the ovary counts of animals with different Complement 5 (Hc) alleles

Chr2:34873343-34951450 bp, - strand -- find the closest markers, look how Dan did it with Lifr and try to follow that.












